lld/ELF/Arch/LoongArch.cpp

```diff
@@ -159,6 +159,10 @@ uint64_t elf::getLoongArchPageDelta(uint64_t dest, uint64_t pc) {
   bool negativeA = lo12(dest) > 0x7ff;
   bool negativeB = (result & 0x8000'0000) != 0;
 
+  // A corner case; directly return the expected result.
+  if (result == 0xfffffffffffff000 && negativeA)
+    return result = 0xffffffff00000000;
```

Comment:
I can see the misbehavior is caused by an overflow in the `result += 0x1000` below; who expected that! Thanks for spotting it.

First of all, stylistically, the `result =` part could be dropped and some `'` could be added to help counting bits; other than that, I don't know if the other arithmetic could overflow too, in which case we might want to guard them too.

Reply:
The current case is raised by some end user. I don't know if there are other cases too. Seems that binutils' processing is in elfnn-loongarch.c (`RELOCATE_CALC_PC64_HI32`); I'm not sure whether it is helpful.

Reply:
Seems that mold uses the same impl as binutils. https://github.com/rui314/mold/blob/v2.3.2/elf/arch-loongarch.cc#L53

Reply:
I had report this bug to mengqinggang and Rui https://github.com/rui314/mold/issues/1131#issuecomment-1771892552
That's because we can not get consistent PC from `R_LARCH_PCALA_HI20` and `R_LARCH_PCALA64_{LO20,HI12}`. With this fix, we will wrong when span (+2 + 4N)G.

```
0x12344ffc orig (pcalau12i, ori, lu32i, lu52i)
0x92345AAA dest
lo12: 0xAAA
pagedelta: 0x80001000
hi20: 80001000
64_lo20: 0
64_hi12: 0
result: 0xffffffff92345AAA
```

Reply:
This special case should probably be added inside `if (negativeA) { ... }` so that it reads less magic.